sstables_loader.hh

```diff
@@ -35,6 +35,7 @@ class view_builder;
 // system. Built on top of the distributed_loader functionality.
 class sstables_loader : public seastar::peering_sharded_service<sstables_loader> {
 public:
+    enum class stream_scope { all, dc, rack, node };
```

Comment:
I don't understand why primary-replica-only is considered to be a sub-kind of possible scopes. The former limits the amount of copies the loaded (and streamed) mutation is sent to, the latter defines if the target node is able to receive it or now

Reply:
Isn't it the same thing?

scope=rack: limit copying to this (or a particular?) rack
scope=datacenter: limit copying to this (or a particular?) datacenter
scope=primary: limit copying to the primary replica

Reply:
> scope=rack: limit copying to this (or a particular?) rack
> scope=datacenter: limit copying to this (or a particular?) datacenter

It's not specified here what is being copied, it can be either single replica or all RF copies of those.

> scope=primary: limit copying to the primary replica

This implies that primary replica can be anywhere, while this PR's cope can say smth like "if primary replica is in another rack, just skip this mutation"

> Isn't it the same thing?

Can be, but I have different interpretation. That is:

scope=X† vs primary_replica_only=false: limit spreading mutations to this X
scope=X† vs primary_replica_only=true: limit spreading primary replicas inside this X

† (X in all, dc, rack, node).

It's worth expanding the list of scopes to be

scope=rack: limit copying to this (or a particular?) rack
scope=primary_replica_in_rack: limit copying primary replica to this (or a particular?) rack
scope=datacenter: limit copying to this (or a particular?) datacenter
scope=primary_replica_in_datacenter: limit copying primary replica to this (or a particular?) datacenter

as @bhalevy suggested above, but it's like opening braces in {scope}X{primary|all} combinations

Reply:
We discussed the option to treat `scope` completely differently and actually name it `filter`:
The target is to have a list of nodes a mutation belongs to. The way to get this list is by _filtering out_ all those that are not in the given dc, rack, node, primary replica or any other filter.

We can supply a few filters to to the mechanism.
Filters are applied orthogonality, one by one. The filter mechanism does not understand that a `node` is included in `dc` or any other understanding of the hierarchical structure of the cluster.

In this way all we need to do is to list the nodes we want to filter and the mutation is sent to everything that is not filtered out.
The advantages are:
- Simplicity.
- We can come up in the future with any filtering mechanism we want, it works the same (e.g. the user gives the exact list of nodes to load-and-stream to)
- We can combine a few filters together.
- `primary-replica` is a filter, combined with the `dc` filter it can send mutations to the primary replica only if it resides in the same dc.

Reply:
> primary-replica is a filter, combined with the dc filter it can send mutations to the primary replica only if it resides in the same dc.

Restore is DC-centric so it doesn't make a lot of sense to filter the primary replica, but rather to have an option to select a primary replica in the datacenter.
After that the admin can repair either the dc, or run a global repair that will replicate the data from the restored datacenter everywhere.

Reply:
> Restore is DC-centric so it doesn't make a lot of sense to filter the primary replica

But it makes sense to filter primary-replica and rack (two filters) - which means it will drop all mutations that their primary replica is not in its rack.
It also make sense to filter primary-replica and node (again, two filters to pass through).
In the future, if we come up with more advanced filters, they will be easily integrated.

Also I agree with @avikivity when he wants [simplicity](https://github.com/scylladb/scylladb/pull/21169#discussion_r1818072013).
For me, understanding `filter`  is simpler than understanding `scopes`.

Reply:
To keep it simple we'll probably need to conflate the Cartesian product of primary replica X scope ro all meaningful combinations to e.g.: --send-to=
- all-replicas
- primary-replica
- all-replicas-in-dc
- primary-replica-in-dc
- all-replicas-in-rack
- primary-replica-in-rack
- this-node
- this-node-if-primary-replica
- this-node-if-primary-replica-in-dc
- this-node-if-peimary-replica-in-rack

Is it indeed simpler than having 2 options?

Reply:
I didn't explained it well, I'll try again:
```c++
future<> load_and_stream(sstring ks_name, sstring cf_name,
            table_id, std::vector<sstables::shared_sstable> sstables,
            bool unlink_sstables, std::vector<stream_filter> filters);
```
where stream_filter is:
```c++
 enum class stream_filter { all, dc, rack, node, primary_replica };
```
And one can provide a list of filters that each one of them is applied independently.
`all` - means empty filter, nothing is filtered.
`primary_replica` means everything is filtered except for the primary-replica.
```c++
// send to primary replica only if it is in the same dc
std::vector<stream_filter> filters = {dc, primary_replica} 
```

Reply:
> To keep it simple we'll probably need to conflate the Cartesian product of primary replica X scope ro all meaningful combinations to e.g.: --send-to=
> 
>  < snipped long list of options >
> 
> Is it indeed simpler than having 2 options?

Thank you, @bhalevy , this is very precise description of the motivation behind this new "scope"

Reply:
> I didn't explained it well, I'll try again:
> 
> ```c++
> future<> load_and_stream(sstring ks_name, sstring cf_name,
>             table_id, std::vector<sstables::shared_sstable> sstables,
>             bool unlink_sstables, std::vector<stream_filter> filters);
> ```
> 
> where stream_filter is:
> 
> ```c++
>  enum class stream_filter { all, dc, rack, node, primary_replica };
> ```
> 
> And one can provide a list of filters that each one of them is applied independently. `all` - means empty filter, nothing is filtered. `primary_replica` means everything is filtered except for the primary-replica.
> 
> ```c++
> // send to primary replica only if it is in the same dc
> std::vector<stream_filter> filters = {dc, primary_replica} 
> ```

Do other filter combinations make sense?
It seems like only primary_replica can be combined with other filters, and so it really makes it an optional that modifies (or restricts) scope.

@xemul All-in-all I think that the `node` scope is redundant.
It's actually how `nodetool refresh` works without the --load-and-stream option.
So I think we can be good with --scope=all (default)|dc|rack plus --primary-replica-only that limits it primary replica in global|dc|rack respectively.

Reply:
> Do other filter combinations make sense?

No. this is only for simplicity and future needs.
After this discussion, it is ok with me have these two gates: promery-replica and scope.

Reply:
let's keep primary-replica-only orthogonal to `scope`, but make it scope-sensitive, so that it would mean primary-replica-in-dc for datacenter scope (see tablet_map::get_primary_replica_within_dc) and primary-replica-in-rack for the rack scope (wold need a getter for that as well).

Reply:
Sorry for rejoining so late.

If this is a filter, let's call it filter.

But I don't understand what filter=dc means. Ignore mutations that don't belong to this datacenter?

Reply:
> Sorry for rejoining so late.
> 
> If this is a filter, let's call it filter.
> 
> But I don't understand what filter=dc means. Ignore mutations that don't belong to this datacenter?

@xemul should answer that.  But AFAIU, scope=dc means you want to stream mutations to the nodes in the local dc that own the token.  For scope=rack, the subset of nodes are the ones in the local rack (typically only one of them will be the owner of each mutation when rf==nr_racks), and with scope=node, mutations are filtered and streamed only to the local node.

Reply:
Then I propose to call it filter=this_node, filter=this_rack, filter=this_datacenter. So it's abundantly clear what it does.

Is scope=node meaningful? with tablets (and with vnode streaming) node token ranges change all the time.

Reply:
> Then I propose to call it filter=this_node, filter=this_rack, filter=this_datacenter. So it's abundantly clear what it does.
> 
> Is scope=node meaningful? with tablets (and with vnode streaming) node token ranges change all the time.

If load balancing is paused then it can be meaningful.
The intended use-case for rack-aware restore is to give all nodes in the rack an inclusive list of sstables that were backed up from this rack and let each node filter the sstables and possibly their content to stream-in only the data it owns.

Reply:
> Then I propose to call it filter=this_node, filter=this_rack, filter=this_datacenter. So it's abundantly clear what it does.

For me `filter` is more confusing than helping: are we _filtering-out_ this-node or are we keeping this-node?
`filter` may have to meanings exclude or include depends on the context.
`scope` has a much more clearer meaning.